### ISSUE TITLE
Compiler bug? / potential hang up issue with vader btl

### ISSUE THREAD

**Issue by t-niemi** [open-mpi/ompi#6014]
### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)

3.0.3, 3.1.3, 4.0.0rc5

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Downloaded source tarball, compiled with gcc 5.4.0 using ./configure --prefix=... & make all install

### Please describe the system on which you are running

* Operating system/version: ubuntu 16.04, 18.04
* Computer hardware: x86_64
* Network type: shared memory

-----------------------------

## Details of the problem

Following the recent updates to vader btl, I tested 3.0.3, 3.1.3 and 4.0.0rc5 to check that the segfault problems are resolved. 3.0.3 seemed to work fine, but for some reason 3.1.3 and 4.0.0rc5 seemed to hang almost immediately in all test cases. I was able to reduce the problem to the following minimal test code:

```c
#include <mpi.h>

int main(int argc, char** argv) {
  int rounds = atoi(argv[1]);
  int i;
  MPI_Init(NULL, NULL);
  
  for (i = 0; i < rounds; i++) {
    MPI_Barrier(MPI_COMM_WORLD);
  }
  
  MPI_Finalize();
}
```

This code reliably hangs when rounds > 100 and np 2 or more.

As far as I understand, 3.0.3, 3.1.3 have pretty much identical code regarding the vader-part. However, for some reason, 3.0.3 works while 3.1.3 does not. I tried to compile a debug version of 3.1.3 with --enable-debug, but that one worked, so the problem could be related to some compiler optimization. 

On another machine I have gcc 4.9.2 and it showed the same behaviour. On a third machine I have gcc 7.3.0, but it seemed to work ok.

As a conclusion, it seems that if older gcc (5.4.0 or 4.9.2) is used in release mode, the vader part can hang up. With newer gcc (eg. 7.3.0) things seem to work.

**Comment by t-niemi** [comment]
I also tried to play around with btl_vader_fbox.h and making the following change

```c
static inline mca_btl_vader_fbox_hdr_t mca_btl_vader_fbox_read_header (mca_btl_vader_fbox_hdr_t *hdr)
{
    /*mca_btl_vader_fbox_hdr_t tmp = {.data_i32 = {.value1 = hdr->data_i32.value1}};;
    opal_atomic_rmb ();
    tmp.data_i32.value0 = hdr->data_i32.value0;*/
    mca_btl_vader_fbox_hdr_t tmp = {.ival = hdr->ival};
    return tmp;
}
```
seemed to resolve the issue with gcc 5.4.0. 

In any case, this is not a big problem as newer gcc versions seem to work and I can personally use openmpi 2.1.1 with the machines that have older gccs. However, this indicates that there can still be potential issues with some compilers / optimizations.

**Comment by hjelmn** [comment]
I will have to take a look at the code generation from gcc 5.4.0. I have used 5.3.0 and 4.9.x without any issue.

FWIW, we plan to address this by implementing an atomic write. This won't help Open MPI v3.x or v4.0.x but will address the issue for newer releases.

**Comment by hjelmn** [comment]
Hmm. Never use one quite that old.  I recommend at least 4.9.x (6.x or newer is even better) even though we official support older versions. I will check 4.8.x today and see if I can reproduce the hang.

**Comment by derbeyn** [comment]
If I'm not wrong, looks like this is the one that is delivered with the rhel 7.6:

[root@openmpi-2 ~]# gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-36)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

[root@openmpi-2 ~]# cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.6 (Maipo)

**Comment by t-niemi** [comment]
I also tested one more compiler and it has the same issue. Below is a list of all my test cases. Each of them is using the system default compiler

4.9.2, Red Hat/Fedora 21, hangs
5.3.1, Red Hat/Fedora 23, hangs
5.4.0, Ubuntu 16.04, hangs
7.3.0, Ubuntu 18.04, ok

**Comment by hjelmn** [comment]
Yeah, not sure why Redhat insists on shipping a product with an obsolete default compiler (RHELL 7.6 is less than a month old). They could at least have upgraded to 4.9.x and had C11 support.

Looking at the issue now.

**Comment by dirtyharrycallahan** [comment]
The default compiler isn't the only one available on Red Hat.

yum install devtoolset-7

source scl_source enable devtoolset-7

which gcc
/opt/rh/devtoolset-7/root/usr/bin/gcc

# gcc --version
gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**Comment by hjelmn** [comment]
Thats good to know that there is a way to get a modern compiler. Problem is the default shipping compiler is very obsolete. Really wish they wouldn't do that. It's like them shipping old flex with RHEL 6. That was a really bad idea (oldflex was very obsolete when RHEL 6.0 was released).

Anyway, I can't get a hang with gcc 4.9.3 with master ...

```
bash$ prun -n 32 -N 32 ./6014 10000
bash$
```

Will try with v4.0.x now.

**Comment by hjelmn** [comment]
Though if it is easy to install gcc 7.x maybe that would make it easier for us to require a newer gcc. The lifespan of RHEL 7 is one of the things that has been a sticking point.

**Comment by hjelmn** [comment]
Master is definitely ok but it could be C11 atomics. 4.0.0 definitely hangs with gcc __sync. Checking without builtin-atomics...

**Comment by hjelmn** [comment]
Fine when configuring with --disable-builtin-atomics.

**Comment by hjelmn** [comment]
@bwbarrett  I am starting to think with older gcc versions ``` __atomic_thread_fence (__ATOMIC_ACQUIRE); ``` is being treated as a no-op on x86_64 with -O2. Looking now.

**Comment by hjelmn** [comment]
Adding:

```c
__asm__ __volatile__("": : :"memory");
```

To the memory barriers makes it work. Looks like a gcc bug maybe? Still have to work around it.

**Comment by hjelmn** [comment]
@t-niemi Can you test with #6033 and verify that it fixes the bug for you?

**Comment by t-niemi** [comment]
@hjelmn I will test it today

**Comment by t-niemi** [comment]
Ok, I tested it with 4.9.2 and 5.3.1 and the patch appears to work (I applied it to 3.1.3). I also tested
--disable-builtin-atomics and that one worked too. 

Googling around brought me to pmix issue https://github.com/pmix/pmix/issues/425, which seems to be relevant.

In any case the occurrence of the issues seems quite random (3.0.3 works, 3.1.3 not) and I'm not sure how new the compiler would need to be, because I had problems with 5.3.1 and 5.4.0, (at least 6, or 7?). But in any case, the patch seems to do the trick.

**Comment by rhc54** [comment]
We will port the patch to PMIx as well since we track the OMPI atomics there - thx for alerting us!

**Comment by jrood-nrel** [comment]
Maybe my issue is different but I think I'm experiencing this as well. Using GCC 6.4.0 and OpenMPI 3.1.3 and running in local shared memory causes my application to hang indefinitely where OpenMPI 3.1.2 (with the vader btl patch) does not hang. Originally in OpenMPI 3.1.2 without the vader patch, I was seeing segfaults that showed up after hours of application runtime, but I haven't tested the vader patch for that case.

**Comment by jsquyres** [comment]
These atomics have just been the gift that keeps on giving... ☹️ 

We have fixed this on master, the v4.0.x branch, and it looks like it's still awaiting merging to the v3.0.x and v3.1.x branches (keep in mind that the Supercomputing trade show is next week -- we're all in panic mode preparing for that...).

**Comment by jrood-nrel** [comment]
Just tested using GCC 7.3.0 and OpenMPI 3.1.3 and my application doesn't hang in that case, so this looks like the same issue. I'm going to continue with this configuration and see if it also solves my segfault problems from OpenMPI 3.1.2.

**Comment by jsquyres** [comment]
@jrood-nrel We have merged the fixes to the v3.0.x and v3.1.x branches, so the latest nightly snapshots should have the fix.  Let us know if this doesn't solve your problem.

https://www.open-mpi.org/nightly/v3.0.x/
https://www.open-mpi.org/nightly/v3.1.x/

**Comment by ersereik** [comment]
> @jrood-nrel We have merged the fixes to the v3.0.x and v3.1.x branches, so the latest nightly snapshots should have the fix. Let us know if this doesn't solve your problem.
> 
> https://www.open-mpi.org/nightly/v3.0.x/
> https://www.open-mpi.org/nightly/v3.1.x/

Hi,

I do not see this fix applied on v3.1.x or latest nightly openmpi-v3.1.x-201812140241-0e27b42.tar.bz2 ?

**Comment by jsquyres** [comment]
This is done and merged to release branches.
